pasture-algorithms/src/normal_estimation.rs

```diff
@@ -0,0 +1,676 @@
+use core::panic;
+
+use kd_tree::{self, KdPoint, KdTree};
+use num_traits::{self};
+use pasture_core::layout::{attributes::POSITION_3D, PointType};
+use pasture_core::{
+    containers::{PerAttributeVecPointStorage, PointBuffer, PointBufferExt},
+    nalgebra::{DMatrix, Vector3},
+};
+
+/// Normal Estimation Algorithm
+/// returns a vector of quintuplets where each quintuplet has the following values: (current point, normal_vector, curvature).
+/// It iterates over all points in the buffer and constructs new point buffers of size k_nn.
+/// Make sure that knn >= 3 holds as it is not possible to construct a plane over less than three point
+///
+/// # Panics
+///
+/// Panics if the number of k nearest neighbors is less than 3 or the point cloud has less than 3 elements.
+///
+/// # Examples
+///
+/// ```
+/// # use kd_tree::{self, KdPoint};
+/// # use pasture_core::nalgebra::Vector3;
+/// # use pasture_core::{containers::InterleavedVecPointStorage, layout::PointType};
+/// # use pasture_algorithms::normal_estimation::compute_normals;
+/// # use pasture_derive::PointType;
+/// # use typenum;
+///
+/// #[repr(C)]
+/// #[derive(PointType, Debug, Clone, Copy)]
+/// struct SimplePoint {
+///     #[pasture(BUILTIN_POSITION_3D)]
+///     pub position: Vector3<f64>,
+///     #[pasture(BUILTIN_INTENSITY)]
+///     pub intensity: u16,
+/// }
+/// impl KdPoint for SimplePoint {
+///     type Scalar = f64;
+///     type Dim = typenum::U3;
+///     fn at(&self, k: usize) -> f64 {
+///         return self.position[k];
+///     }
+/// }
+/// fn main() {
+///     let points = vec![
+///         SimplePoint {
+///             position: Vector3::new(11.0, 22.0, 154.0),
+///             intensity: 42,
+///         },
+///         SimplePoint {
+///             position: Vector3::new(12.0, 23.0, 0.0),
+///             intensity: 84,
+///         },
+///         SimplePoint {
+///             position: Vector3::new(103.0, 84.0, 2.0),
+///             intensity: 84,
+///         },
+///         SimplePoint {
+///             position: Vector3::new(101.0, 0.0, 1.0),
+///             intensity: 84,
+///         },
+///     ];
+///
+///     let mut interleaved = InterleavedVecPointStorage::new(SimplePoint::layout());
+///
+///     interleaved.push_points(points.as_slice());
+///
+///     let solution_vec = compute_normals::<InterleavedVecPointStorage, SimplePoint>(&interleaved, 4);
+///     for solution in solution_vec {
+///        println!(
+///             "Point: {:?}, n_x: {}, n_y: {}, n_z: {}, curvature: {}",
+///            solution.0, solution.1, solution.2, solution.3, solution.4
+///         );
+///     }
+/// }
+/// ```
+
+pub fn compute_normals<T: PointBuffer, P: PointType + KdPoint + Copy>(
```

Comment:
I just realized that you are passing the point type as another generic parameter. If you do it this way, you have to enforce that `P::layout() == *point_cloud.point_layout()`! It would be better if you didn't need the generic parameter `P` though. You can just enforce that the `PointLayout` of `point_cloud` has the `POSITION_3D` attribute and then instead of doing `point_cloud.iter_point()` you do `point_cloud.iter_attribute_as::<Vector3<f64>>(&POSITION_3D)` to only get the positions. Currently, afaik the `kd_tree` create does not implement `KdPoint` for the `nalgebra` types (e.g. `Vector3<T>`), so you have to convert the positions to `[f64;3]`. 

The advantage of this change would be that the method will work with any point cloud, even if the underlying type is not known at compile-time. Think about reading a point cloud from a `LAS` file: The resulting `PointLayout` will be determined at runtime, so you can't know the type for `P` at compile-time.